signalflow/client.go

```diff
@@ -0,0 +1,326 @@
+		return err
+	}
+}
+
+// AccessToken can be used to provide a SignalFx organization access token or
+// user access token to the SignalFlow client.
+func AccessToken(token string) ClientParam {
+	return func(c *Client) error {
+		c.token = token
+		return nil
+	}
+}
+
+// UserAgent allows setting the `userAgent` field when authenticating to
+// SignalFlow.  This can be useful for accounting how many jobs are started
+// from each client.
+func UserAgent(userAgent string) ClientParam {
+	return func(c *Client) error {
+		c.userAgent = userAgent
+		return nil
+	}
+}
+
+// MetadataTimeout is the default amount of time that calls to metadata
+// accessors on a SignalFlow Computation instance will wait to receive the
+// metadata from the backend before failing and returning a zero value. Usually
+// metadata comes in very quickly from the stream after the job start.
+func MetadataTimeout(timeout time.Duration) ClientParam {
+	return func(c *Client) error {
+		c.defaultMetadataTimeout = timeout
+		return nil
+	}
+}
+
+// ReadTimeout sets the duration to wait between messages that come on the
+// websocket.  If the resolution of the job is very low, this should be
+// increased.
+func ReadTimeout(timeout time.Duration) ClientParam {
+	return func(c *Client) error {
+		c.readTimeout = timeout
+		return nil
+	}
+}
+
+// NewClient makes a new, but uninitialized, SignalFlow client.
+func NewClient(options ...ClientParam) (*Client, error) {
+	c := &Client{
+		streamURL: &url.URL{
+			Scheme: "wss",
+			Host:   "stream.us0.signalfx.com",
+			Path:   "/v2/signalflow",
+		},
+		readTimeout:            1 * time.Minute,
+		writeTimeout:           5 * time.Second,
+		outgoingMessages:       make(chan interface{}),
+		channelsByName:         make(map[string]*Channel),
+		defaultMetadataTimeout: 5 * time.Second,
+	}
+
+	for i := range options {
+		if err := options[i](c); err != nil {
+			return nil, err
+		}
+	}
+
+	return c, nil
+}
+
+func (c *Client) ensureInitialized() error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	var err error
+	if !c.isInitialized {
+		err = c.initialize()
+		// The mutex also acts as a memory barrier to ensure this write will be
+		// seen by any goroutine that obtains the lock after the last Unlock,
+		// see https://golang.org/ref/mem#tmp_8.
+		c.isInitialized = true
+	}
+	return err
+}
+
+// Assumes c.Mutex is held when called.  Gets the client connection in a state
+// that is ready for execute requests.
+func (c *Client) initialize() error {
+	authenticatedCond := sync.NewCond(&c.lock)
+
+	if c.isShutdown {
+		return errors.New("cannot initialize client after shutdown")
+	}
+
+	c.ctx, c.cancel = context.WithCancel(context.Background())
+
+	if c.conn == nil {
+		var err error
+		c.conn, err = connect(c.ctx, c.streamURL)
+		if err != nil {
+			return err
+		}
+
+		go c.keepWritingMessages()
+		go c.keepReadingMessages(authenticatedCond)
+		c.authenticate()
```

Comment:
I had to go trace through this to grok that we had to start the goroutines before calling this to actually do the writing. Might be worth explaining the order as a comment for future spelunkers?